packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -178,21 +158,16 @@ export class PendingStateManager implements IDisposable {
         this._pendingMessagesCount++;
     }
 
-    /**
-     * Called when the FlushMode is updated. Adds the FlushMode to the pending state queue.
-     * @param flushMode - The flushMode that was updated.
-     */
-    public onFlushModeUpdated(flushMode: FlushMode) {
-        this.pendingStates.push({ type: "flushMode", flushMode });
-    }
-
     /**
      * Called when flush() is called on the ContainerRuntime to manually flush messages.
+     * @param isImmediateBatch - add the "flush" message to the pending states even when flushMode is Immediate
      */
-    public onFlush() {
-        // If the FlushMode is Immediate, we don't need to track an explicit flush call because every message is
-        // automatically flushed. So, flush is a no-op.
-        if (this.stateHandler.flushMode() === FlushMode.Immediate) {
+    public onFlush(isImmediateBatch: boolean = false) {
+        /**
+         * If the FlushMode is Immediate, we don't need to track an explicit flush call because every message is
+         * automatically flushed. So, flush is a no-op.
+         */
+        if (!isImmediateBatch && this.flushMode === FlushMode.Immediate) {
```

Comment:
Can this check be moved out to container runtime layer? It feels like this layer (if possible) should just track batches and ops and not care about flush modes. I.e. it's responsibility of container runtime layer to convert its API surface to a stream of batches and ops (illumining flattening nested orderSequentially calls)

Reply:
Maybe it's for another set of refactoring :)

Reply:
[AB#1107](https://dev.azure.com/fluidframework/internal/_workitems/edit/1107)